### DISCUSSION TITLE
Using lidar sensor in orbit

### DISCUSSION THREAD

**Discussion by LukasVindbjerg** [isaac-sim/IsaacLab#291] [Q&A] [answered]
Hi, 

I have a robot that uses a 2D lidar sensor. I have added the RTX Lidar sensor to the robot inside Isaac Sim. When I run the simulation from orbit I can see the dots created by the lidar sensors, indicating that it is working. 

However, I'm not sure how to access the data generated by the lidar sensors from orbit. From the documentation it reads:
"USD-prim sensors: Available in Omniverse and require creating a USD prim for them. For instance, RTX ray tracing camera and lidar sensors."

I'm not sure if this means that I can access the lidar sensor I created in Isaac Sim from its prim path, or if I have to spawn a new one from the orbit API. Can someone clarify the workflow for accessing data from lidar sensors in orbit?

**Comment by Mayankm96** [comment]
Hi @LukasVindbjerg ,

We had planned to add support for RTX lidar but since it wasn't parallelized, we decided to remove it in favor of our custom wrap-based ray-caster. However, there are certain benefits to the RTX implementation (such as surface reflectivity and realism beyond physical meshes).

At this point, the feature is not in our immediate development plans. However, it would be great to support this feature. As a starting point, you can make the sensor class similar to how `Camera` works in Orbit. If you need more guidance, please do not hesitate to ask.

**Reply by LukasVindbjerg** [reply]
Thank you for your answer. Can I achieve a similar 2D lidar behavior by implementing a custom circular pattern for the RayCaster sensor?

**Reply by LukasVindbjerg** [reply]
Hi @Mayankm96 

I have been working on a `Lidar` class similar to the current `Camera` class. I think that the implementation is close to working, but I'm missing the last piece which is actually getting the lidar output from the simulation. 

I found the following part of the code to be a bit hard to translate:

```python
        for lidar_prim_path in self._view.prim_paths:
            lidar_prim = omni.usd.get_context().get_stage().GetPrimAtPath(lidar_prim_path)

            self._sensor_prims.append(lidar_prim)

            render_prod_path = rep.create.render_product(lidar_prim_path, resolution=[1, 1])
            
            
            for name in self.cfg.data_types:
            
                rep_annotator = rep.AnnotatorRegistry.get_annotator(name, device=device_name)
                
                rep_annotator.attach(render_prod_path)
            
                self._rep_registry[name].append(rep_annotator)
```

The tricky part is the `name` variable field for the `get_annotator()`. The stack trace provided a list of different fields to choose from:

```python
    data_types: list[str] = ["pointcloud"]
    
    """
  Available annotators: ['camera_params', 'rgb', 'normals', 'motion_vectors', 'cross_correspondence', 'occlusion',
   'distance_to_image_plane', 'distance_to_camera', 'LdrColor', 'HdrColor', 'SmoothNormal', 'BumpNormal', 'AmbientOcclusion', 
   'Motion2d', 'DiffuseAlbedo', 'SpecularAlbedo', 'Roughness', 'DirectDiffuse', 'DirectSpecular', 'Reflections', 'IndirectDiffuse', 
   'DepthLinearized', 'EmissionAndForegroundMask', 'PtDirectIllumation', 'PtGlobalIllumination', 'PtReflections', 'PtRefractions', 
   'PtSelfIllumination', 'PtBackground', 'PtWorldNormal', 'PtWorldPos', 'PtZDepth', 'PtVolumes', 'PtDiffuseFilter', 'PtReflectionFilter', 
   'PtRefractionFilter', 'PtMultiMatte0', 'PtMultiMatte1', 'PtMultiMatte2', 'PtMultiMatte3', 'PtMultiMatte4', 'PtMultiMatte5', 'PtMultiMatte6', 
   'PtMultiMatte7', 'primPaths', 'bounding_box_2d_tight_fast', 'bounding_box_2d_tight', 'bounding_box_2d_loose_fast', 
   'bounding_box_2d_loose', 'bounding_box_3d_360', 'bounding_box_3d_fast', 'bounding_box_3d', 'semantic_segmentation', 
   'instance_segmentation_fast', 'instance_segmentation', 'CameraParams', 'skeleton_data', 'pointcloud', 'CrossCorrespondence', 
   'MotionVectors', 'IsaacNoop', 'IsaacReadCameraInfo', 'IsaacReadTimes', 'IsaacReadSimulationTime', 
   'LdrColorSDIsaacConvertRGBAToRGB', 'DistanceToImagePlaneSDIsaacConvertDepthToPointCloud', 
   'RtxSensorCpuIsaacReadRTXLidarData', 'RtxSensorCpuIsaacComputeRTXLidarPointCloud', 
   'RtxSensorCpuIsaacCreateRTXLidarScanBuffer', 'RtxSensorCpuIsaacComputeRTXLidarFlatScan', 
   'RtxSensorCpuIsaacComputeRTXRadarPointCloud']
   
    """
```
I have tried a bunch of different RTX ones, but all of the fields in the `output` variable are empty. Do you have any suggestions as to how I can set up the annotator correctly so that the `output` gets populated?

**Reply by Mayankm96** [reply]
Wondering if it is one of the known issues that it takes a few rendering steps for the buffers to fill up?

**Reply by Mayankm96** [reply]
@pascal-roth Can you help out here since you already did some version of this, if I remember correctly? Thank you!

**Reply by LukasVindbjerg** [reply]
Found a solution for getting the lidar data populated now. It works perfectly when I'm running in non-headless mode. However, if I run the training in headless mode all of the lidar values become 0. Have you encountered a similar problem before with the raycaster or camera?

**Reply by Mayankm96** [reply]
Probably need to provide `--offscreen_render` flag to allow rendering when in headless mode.

> https://isaac-orbit.github.io/orbit/source/tutorials/00_sim/launch_app.html#understanding-the-output-of-help

**Reply by LukasVindbjerg** [reply]
Thanks @Mayankm96, that almost solved the issue. However, now the lidar measurements seem to be updated only in the first frame, and then they have the same value for the rest of the training when I use `--offscreen_render`. E.g.:

```
lidar ranges: tensor([[30.0000, 30.0000, 30.0000,  ..., 30.0000, 17.6805, 30.0000],
        [17.5025, 30.0000, 17.6963,  ..., 30.0000, 17.6805, 30.0000],
        [17.5025, 30.0000, 17.6963,  ..., 30.0000, 17.6805, 30.0000],
        ...,
        [30.0000, 30.0000, 30.0000,  ..., 30.0000, 30.0000, 30.0000],
        [30.0000, 30.0000, 30.0000,  ..., 30.0000, 30.0000, 30.0000],
        [30.0000, 30.0000, 30.0000,  ..., 30.0000, 30.0000, 30.0000]],
       device='cuda:0')
```
It works as expected when I run on non-headless mode, e.g.:

```
lidar ranges: tensor([[0.6031, 0.5970, 0.5911,  ..., 1.4989, 1.5016, 1.5043],
        [0.3798, 0.3791, 0.3783,  ..., 2.9556, 2.9827, 3.0104],
        [0.3971, 0.3975, 0.3979,  ..., 1.6996, 1.6691, 1.6390],
        [0.3195, 0.3188, 0.3181,  ..., 2.9194, 2.9439, 2.9689]],
       device='cuda:0')
```
I'm not sure if this is due to an issue with my implementation of the lidar class. Have you encountered similar issues with the `--offscreen_render` before?

**Reply by LukasVindbjerg** [reply]
It works when I use the combination of `--headless`, `--offscreen_render`, and `--video`. However, this is quite expensive in terms of memory, so I would really like to find a solution that avoids using the `--video` flag.

**Reply by Mayankm96** [reply, marked as answer]
You need to call `env.render()` to allow rendering of sensor data. I noticed that we right now only call it implicitly in the `step` call when GUI is enabled. When you use RSL-RL or RecordVideo wrapper, they internally call the `render` after step so you see it as working.

> https://github.com/NVIDIA-Omniverse/orbit/blob/main/source/extensions/omni.isaac.orbit/omni/isaac/orbit/envs/rl_task_env.py#L175-L177

But I'm not happy with this arrangement. Would be nice to have a general way to "call" render when a sensor needs it.

**Reply by LukasVindbjerg** [reply]
Thank you very much for your help! I changed the line to:
```python
if self.sim.has_gui() or self.sim._offscreen_render:
    self.sim.render()
```
Probably not the most ideal solution - but in my case I can now simulate multiple environments with lidar, which was the goal.

While the code is far from polished I'll leave a link to my fork here in case anyone else is interested in doing something similar with lidar sensors:
https://github.com/Lukas-Malthe-MSc/orbit/tree/f1tenth-environment

**Reply by Mayankm96** [reply]
Nice! Would you like to contribute the Lidar sensor implementation to Orbit? @pascal-roth for vis.

**Reply by LukasVindbjerg** [reply]
Yes absolutely. However, right now the code has just been implemented to suit my own needs and could use a bit more work put into it. Also, I recently tried to pull the changes from the main branch into my development branch - which unfortunately seemed to break some of the Isaac Sim APIs that I use. 

I'd be happy to try and fix the lidar implementation. But right now other parts of my project are a higher priority for me.

**Reply by Mayankm96** [reply]
Just for an update. We have fixed the behavior in this commit: 7ac5588c12406a323772b2c66f72ff4db546ebac
